src/parallel/task_scheduler.cpp

```diff
@@ -78,22 +83,39 @@ struct ConcurrentQueue {
 
 void ConcurrentQueue::Enqueue(ProducerToken &token, shared_ptr<Task> task) {
 	lock_guard<mutex> lock(qlock);
-	q.push(std::move(task));
+	QueueProducerToken *token_addr = token.token.get();
+	q[token_addr].push(std::move(task));
 }
 
 bool ConcurrentQueue::DequeueFromProducer(ProducerToken &token, shared_ptr<Task> &task) {
 	lock_guard<mutex> lock(qlock);
 	if (q.empty()) {
 		return false;
 	}
-	task = std::move(q.front());
-	q.pop();
+
+	QueueProducerToken *token_addr = token.token.get();
+
+	const auto it = q.find(token_addr);
+	if (it == q.end()) {
+		return false;
```

Comment:
Should this ever happen? Perhaps this should be an internal exception?

Reply:
Could this be the case, if all tasks of a query are currently blocked and we are still stuck here
```c++
while (!IsFinishedOrBlocked(ExecuteTaskInternal(lock))) {
}
```

Reply:
The token should still be active in this case, no?

Reply:
That's true